src/common/health.rs

```diff
@@ -14,260 +14,209 @@ use collection::shards::CollectionId;
-        if self.is_running().await {
-            log::warn!("Ready::spawn_task: ReadyTask is already running");
-            return;
-        }
-
-        let Some(mut task) = self.try_cleanup_task().await else {
-            return;
-        };
+    async fn wait_ready(&self) -> bool {
+        let is_ready_signal = self.is_ready_signal.notified();
 
         if self.is_ready() {
-            log::error!("TODO");
-            return;
-        }
-
-        *task = Some(self.runtime.spawn(self.task().exec()));
-    }
-
-    async fn try_cleanup_task(
-        &self,
-    ) -> Option<sync::RwLockWriteGuard<'_, Option<task::JoinHandle<()>>>> {
-        let mut task = self.task.write().await;
-
-        if is_task_running(task.as_ref()) {
-            log::debug!("Ready::try_cleanup_task: ReadyTask is running");
-            return None;
-        }
-
-        let result = match task.take() {
-            Some(task) => task.await,
-            None => Ok(()),
-        };
-
-        if let Err(err) = result {
-            log::error!("ReadyTask failed: {err}");
+            return true;
         }
 
-        Some(task)
-    }
-
-    async fn state(&self) -> (bool, bool) {
-        task_state(self.task.read().await.as_ref())
-    }
-
-    async fn is_running(&self) -> bool {
-        is_task_running(self.task.read().await.as_ref())
-    }
-
-    fn task(&self) -> Task {
-        Task {
-            is_ready: self.is_ready.clone(),
-            notify_ready: self.notify_ready.clone(),
-            toc: self.toc.clone(),
-            consensus_state: self.consensus_state.clone(),
-            check_ready: self.check_ready.subscribe(),
-        }
+        time::timeout(DEFAULT_READY_CHECK_TIMEOUT, is_ready_signal)
+            .await
+            .is_ok()
     }
 }
 
-fn task_state<T>(task: Option<&task::JoinHandle<T>>) -> (bool, bool) {
-    (is_task_running(task), is_task_spawned(task))
-}
-
-fn is_task_running<T>(task: Option<&task::JoinHandle<T>>) -> bool {
-    task.map_or(false, |task| !task.is_finished())
-}
-
-fn is_task_spawned<T>(task: Option<&task::JoinHandle<T>>) -> bool {
-    task.is_some()
-}
-
 pub struct Task {
-    is_ready: Arc<AtomicBool>,
-    notify_ready: Arc<sync::Notify>,
     toc: Arc<TableOfContent>,
     consensus_state: ConsensusStateRef,
-    check_ready: broadcast::Receiver<()>,
+    is_ready: Arc<AtomicBool>,
+    is_ready_signal: Arc<sync::Notify>,
+    check_ready_signal: Arc<sync::Notify>,
+    cancel: cancel::CancellationToken,
 }
 
 impl Task {
     pub async fn exec(mut self) {
-        if self.check_shortcut() {
-            return;
+        while self.exec_catch_unwind().await.is_err() {
+            log::error!("health::Task panicked") // TODO!?
         }
+    }
 
-        let mut commit_to_await = u64::MAX;
+    async fn exec_catch_unwind(&mut self) -> thread::Result<()> {
+        panic::AssertUnwindSafe(self.exec_cancel())
+            .catch_unwind()
+            .await
+    }
```

Comment:
IMO, this is simpler than the previous `spawn_task()` implementation. The only way it can fail, is if `log::error!(...)` panics.

Reply:
**UPD:** if `downcast_str` or `log::error!(...)` panics